Add intercom construction/deconstruction

Fre3bie opened the pull request:
:cl: Fre3bie
rscadd: Added construction and deconstruction of intercoms
/:cl:

Fre3bie committed:
Add intercom construction/deconstruction

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -11,6 +11,8 @@
 	cell = null
 	power_usage = 0
+	var/wiresexposed = FALSE
+	var/buildstage = 2 //2 = wired/built, 1 = circuit installed, 0 = frame
```
dmdoc

```suggestion
	/// 2 = wired/built, 1 = circuit installed, 0 = frame
	var/buildstage = 2 
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -138,13 +140,29 @@
 	STOP_PROCESSING(SSobj, src)
 	return ..()
 
+/obj/item/device/radio/intercom/New(var/loc, var/dir, atom/frame)
```
`Initialize()` is preferred unless you absolutely must use `New()`

Also, `var/` is redundant in proc parameters.

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+/obj/item/device/radio/intercom/attackby(obj/item/W as obj, mob/user as mob)
+	switch (buildstage)
+		if (2)
+			..()
```
`attackby()` probably shouldn't be doing other interactions if the conditions below were met and processed - This is how you get weird, unexpected results, double feedback messages that don't make sense, etc.

Fre3bie replied:
I included it there because it calls the super function for allowing hacking and toggling attachments, which should only be possible during the specific build state. I tried placing it elsewhere however it resulted in it not calling properly and losing that functionality for the intercoms. 

If there's another way of getting around it I'd be happy to make the changes, but I wasn't able to get it working properly otherwise.

SierraKomodo replied:
I'll have to do some experimenting when I'm home but it should still work if you move the `..()` call after the wirecutter handling in this section.

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+			..()
+			if (isScrewdriver(W))
+				wiresexposed = !wiresexposed
+				to_chat(user, "The wires have been [wiresexposed ? "exposed" : "unexposed"]")
```
```suggestion
				to_chat(user, SPAN_NOTICE("The wires have been [wiresexposed ? "exposed" : "unexposed"]"))
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+				update_icon()
+				return
+			else if (wiresexposed && isWirecutter(W))
+				user.visible_message("<span class='warning'>[user] has cut the wires inside \the [src]!</span>", "You have cut the wires inside \the [src].")
```
```suggestion
				user.visible_message(
					SPAN_WARNING("\The [user] has cut the wires inside \the [src]!"),
					SPAN_NOTICE("You have cut the wires inside \the [src].")
				)
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+			if (isCoil(W))
+				var/obj/item/stack/cable_coil/C = W
+				if (C.use(5))
+					to_chat(user, "<span class='notice'>You wire \the [src].</span>")
```
```suggestion
					to_chat(user, SPAN_NOTICE("You wire \the [src]."))
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+					return
+
+				else
+					to_chat(user, "<span class='warning'>You need 5 pieces of cable to wire \the [src].</span>")
```
```suggestion
					to_chat(user, SPAN_WARNING("You need 5 pieces of cable to wire \the [src]."))
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+					return
+
+			else if (isCrowbar(W))
+				to_chat(user, "You start prying out the circuits.")
```
```suggestion
				to_chat(user, SPAN_NOTICE("You start prying out the circuits."))
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+
+			else if (isCrowbar(W))
+				to_chat(user, "You start prying out the circuits.")
+				playsound(src.loc, 'sound/items/Crowbar.ogg', 50, 1)
```
```suggestion
				playsound(loc, 'sound/items/Crowbar.ogg', 50, 1)
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+				return
+			else if (wiresexposed && isWirecutter(W))
+				user.visible_message("<span class='warning'>[user] has cut the wires inside \the [src]!</span>", "You have cut the wires inside \the [src].")
+				playsound(src.loc, 'sound/items/Wirecutter.ogg', 50, 1)
```
```suggestion
				playsound(loc, 'sound/items/Wirecutter.ogg', 50, 1)
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+				to_chat(user, "You start prying out the circuits.")
+				playsound(src.loc, 'sound/items/Crowbar.ogg', 50, 1)
+				if (do_after(user, 20) && buildstage == 1)
+					to_chat(user, "You pry out the circuit!")
```
```suggestion
					to_chat(user, SPAN_NOTICE("You pry out the circuit!"))
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+				playsound(src.loc, 'sound/items/Crowbar.ogg', 50, 1)
+				if (do_after(user, 20) && buildstage == 1)
+					to_chat(user, "You pry out the circuit!")
+					var/obj/item/intercom_electronics/circuit = new /obj/item/intercom_electronics()
+					circuit.dropInto(user.loc)
```
```suggestion
					new /obj/item/intercom_electronics(get_turf(user))
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
 
 	return canhear_range
 
+/obj/item/device/radio/intercom/attackby(obj/item/W as obj, mob/user as mob)
```
I don't believe the `as XYZ` is necessary here since there are already type paths being assigned to these vars.
```suggestion
/obj/item/device/radio/intercom/attackby(obj/item/W, mob/user)
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+				return
+		if (0)
+			if (istype(W, /obj/item/intercom_electronics))
+				to_chat(user, "You insert the circuit!")
```
```suggestion
				to_chat(user, SPAN_NOTICE("You insert the circuit!"))
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+				update_icon()
+				return
+			else if (isWrench(W))
+				to_chat(user, "You remove the intercom assembly from the wall!")
```
```suggestion
				to_chat(user, SPAN_NOTICE("You remove the intercom assembly from the wall!"))
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+			else if (isWrench(W))
+				to_chat(user, "You remove the intercom assembly from the wall!")
+				new /obj/item/frame/intercom(get_turf(user))
+				playsound(src.loc, 'sound/items/Ratchet.ogg', 50, 1)
```
```suggestion
				playsound(loc, 'sound/items/Ratchet.ogg', 50, 1)
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -161,7 +179,66 @@
+
 /obj/item/device/radio/intercom/Process()
+	if (wiresexposed)
+		on = 0
```
```suggestion
		on = FALSE
```

SierraKomodo reviewed (changes requested):
I also feel most of the `to_chat()` calls should probably be `visible_message()` instead.

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -90,9 +91,21 @@
 	START_PROCESSING(SSobj, src)
+
+	if (dir)
+		src.set_dir(dir)
```
```suggestion
		set_dir(dir)
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -90,9 +91,21 @@
+		wiresexposed = TRUE
+		pixel_x = (dir & 3) ? 0 : (dir == 4 ? -21 : 21)
+		pixel_y = (dir & 3) ? (dir == 1 ? -28 : 23) : 0
+		update_icon()
```
You're already calling `update_icon()` right after this if block.

```suggestion
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -179,18 +178,23 @@
 			if (isScrewdriver(W))
 				wiresexposed = !wiresexposed
-				to_chat(user, "The wires have been [wiresexposed ? "exposed" : "unexposed"]")
+				user.visible_message(SPAN_NOTICE("The wires have been [wiresexposed ? "exposed" : "unexposed"]"))
```
Visible messages should also be adjusted to make sense for a visible message - Describe who, what, and where.

A `visible_message()` is a message displayed to everyone in range to see the object or mob calling the proc, while `to_chat()` is only displayed to the specific mob provided in the proc call.

```suggestion
				user.visible_message(SPAN_NOTICE("\The [user] has [wiresexposed ? "exposed" : "unexposed"] \the [src]'s wiring."))
```

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -201,54 +205,53 @@
 				var/obj/item/stack/cable_coil/C = W
 				if (C.use(5))
-					to_chat(user, "<span class='notice'>You wire \the [src].</span>")
+					user.visible_message(SPAN_NOTICE("You wire \the [src]."))
```
See above

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -201,54 +205,53 @@
 
 				else
-					to_chat(user, "<span class='warning'>You need 5 pieces of cable to wire \the [src].</span>")
+					user.visible_message(SPAN_WARNING("You need 5 pieces of cable to wire \the [src]."))
```
This one can stay a to_chat - It's a user feedback message and it's not affecting the state of the intercom.

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -201,54 +205,53 @@
 			else if (isCrowbar(W))
-				to_chat(user, "You start prying out the circuits.")
-				playsound(src.loc, 'sound/items/Crowbar.ogg', 50, 1)
+				user.visible_message(SPAN_NOTICE("You start prying out the circuits."))
```
See above about visible messages

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -201,54 +205,53 @@
-					to_chat(user, "You pry out the circuit!")
-					var/obj/item/intercom_electronics/circuit = new /obj/item/intercom_electronics()
-					circuit.dropInto(user.loc)
+					user.visible_message(SPAN_NOTICE("You pry out the circuit!"))
```
See above about visible messages

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -201,54 +205,53 @@
 		if (0)
 			if (istype(W, /obj/item/intercom_electronics))
-				to_chat(user, "You insert the circuit!")
+				user.visible_message(SPAN_NOTICE("You insert the circuit!"))
```
See above about visible messages

SierraKomodo commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -201,54 +205,53 @@
 				return
 			else if (isWrench(W))
-				to_chat(user, "You remove the intercom assembly from the wall!")
+				user.visible_message(SPAN_NOTICE("You remove the intercom assembly from the wall!"))
```
See above about visible messages

SierraKomodo reviewed (commented):
Clicking the intercom with tools opens up the nanoui for the intercom - Clicking it with a screwdriver results in opening the main window _and_ the wiring panel window, then the main window closing itself.

This is technically pre-existing behaviour but should probably be fixed here because it'll get in the way of construction/deconstruction.

SierraKomodo commented:
> Clicking the intercom with tools opens up the nanoui for the intercom - Clicking it with a screwdriver results in opening the main window _and_ the wiring panel window, then the main window closing itself.
> 
> This is technically pre-existing behaviour but should probably be fixed here because it'll get in the way of construction/deconstruction.

Actually this is likely a side effect of the `..()` call before tool interactions mentioned abive.

SierraKomodo reviewed (changes requested):
Construction/deconstruction works aside from the `..()` oddity, but could use some QOL tweaks too. Namely, there needs to be ways to identify what construct state the intercom is - Preferably this will be with a unique sprite for each state (One that's empty, one that shows the circuit installed, one that shows the wiring installed), but can also be done through `examine()` text.

Fre3bie commented:
I've adjusted `radio.dm` where it was calling which interface to open, and added a check whether the radio is in the correct state to be showing the wires or the main interface. This fixes the issue mentioned and now only opens the wire interface when using a screwdriver on the intercoms. Additionally the sprites for each build state have been added, originally I avoided it because I was concerned about extra bloat of sprites, but understand it's better to know what state it's in at a glance rather than using trial and error with your tools to see what works at the time.

SierraKomodo reviewed (commented):
Not going to be blocking but would you be willing to also add codex entries for construction/deconstruction steps to the intercom, intercom frame, and intercom circuit? These can be added by overriding the `get_mechanics_info()` proc and returning the information in HTML format, i.e., for the intercom itself:

```dm
/obj/item/device/radio/intercom/get_mechanics_info()
	. = ..()
	. += "<p>To deconstruct:</p>\
		<ol>
			<li>Screwdriver to open the panel</li>
			<li>Wirecutters to remove the wiring</li>
			<li>Crowbar to remove the circuitboard</li>
			<li>Wrench to remove the frame from the wall</li>
		</ol>"
```

Fre3bie commented:
> would you be willing to also add codex entries for construction/deconstruction steps to the intercom, intercom frame, and intercom circuit?

Added the codex entries for the three items and squashed.
